File: ising-sim/src/sim/session.ts
import type { GeometryName } from "../config/geometries";
import { initParamsToConfig, type SimConfig, type SimInitParams } from "./config";
import { SimulationError } from "./errors";
import { GeometryOrchestrator, type GeometryInstance } from "./geometry";
import { NearestNeighbor } from "./interaction";
import { Lattice } from "./lattice";
import { energy, magnetization } from "./metrics";
import { acceptanceRate, sweep, type SweepStats } from "./monteCarlo";
import { SeededRng, sessionRngSeed } from "./rng";

export interface SessionSnapshot {
  spins: number[];
  dimensions: readonly number[];
  rank: number;
  maxNeighbors: number;
  step: number;
  geometry: GeometryName;
}

export interface SessionMetrics {
  energy: number;
  magnetization: number;
  acceptanceRate: number | null;
  temperature: number;
  field: number;
  coupling: number;
}

export class SimulationSession {
  private readonly layout: GeometryInstance;
  private interaction: NearestNeighbor;
  private readonly lattice: Lattice;
  private config: SimConfig;
  private readonly rng: SeededRng;
  private step = 0;
  private running = false;
  private sweepsPerTick = 1;
  private lastAcceptanceRate: number | null = null;
  private cachedSnapshot: SessionSnapshot;

  private constructor(
    layout: GeometryInstance,
    interaction: NearestNeighbor,
    lattice: Lattice,
    config: SimConfig,
    rng: SeededRng,
  ) {
    this.layout = layout;
    this.interaction = interaction;
    this.lattice = lattice;
    this.config = config;
    this.rng = rng;
    this.cachedSnapshot = {
      spins: [],
      dimensions: [...layout.dimensions],
      rank: layout.definition.rank,
      maxNeighbors: layout.maxNeighbors(),
      step: 0,
      geometry: layout.definition.name,
    };
    this.refreshSnapshot();
  }

  static create(params: SimInitParams = {}): SimulationSession {
    const config = initParamsToConfig(params);
    const layout = GeometryOrchestrator.create(config.geometry, config.dimensions);
    const interaction = new NearestNeighbor(config.coupling);
    const lattice = Lattice.create(layout.numSites(), config.seed);
    const rngSeed = sessionRngSeed(config.seed);
    const rng =
      rngSeed === undefined
        ? new SeededRng(Date.now() >>> 0)
        : SeededRng.fromSeed(rngSeed);

    return new SimulationSession(layout, interaction, lattice, config, rng);
  }

  start(stepsPerTick = 1): void {
    this.sweepsPerTick = Math.max(1, stepsPerTick);
    this.running = true;
  }

  pause(): void {
    this.running = false;
  }

  isRunning(): boolean {
    return this.running;
  }

  getSweepsPerTick(): number {
    return this.sweepsPerTick;
  }

  setParams(params: {
    temperature?: number;
    field?: number;
    coupling?: number;
  }): void {
    if (params.temperature !== undefined) {
      if (params.temperature <= 0) {
        throw SimulationError.invalidTemperature();
      }
      this.config.temperature = params.temperature;
    }
    if (params.field !== undefined) {
      this.config.field = params.field;
    }
    if (params.coupling !== undefined) {
      this.config.coupling = params.coupling;
      this.interaction = new NearestNeighbor(params.coupling);
    }
  }

  runSweeps(sweeps: number): SweepStats {
    if (sweeps <= 0) {
      throw SimulationError.invalidStep(sweeps);
    }

    const total: SweepStats = { attempts: 0, acceptances: 0 };
    for (let sweepIndex = 0; sweepIndex < sweeps; sweepIndex += 1) {
      const stats = sweep(
        this.lattice,
        this.layout.core,
        this.interaction,
        this.config,
        this.rng,
      );
      total.attempts += stats.attempts;
      total.acceptances += stats.acceptances;
      this.step += 1;
    }

    this.lastAcceptanceRate = acceptanceRate(total);
    return total;
  }

  getSnapshot(): SessionSnapshot {
    this.refreshSnapshot();
    return {
      spins: [...this.cachedSnapshot.spins],
      dimensions: [...this.cachedSnapshot.dimensions],
      rank: this.cachedSnapshot.rank,
      maxNeighbors: this.cachedSnapshot.maxNeighbors,
      step: this.cachedSnapshot.step,
      geometry: this.cachedSnapshot.geometry,
    };
  }

  getMetrics(): SessionMetrics {
    return {
      energy: energy(this.lattice, this.layout.core, this.interaction, this.config),
      magnetization: magnetization(this.lattice),
      acceptanceRate: this.lastAcceptanceRate,
      temperature: this.config.temperature,
      field: this.config.field,
      coupling: this.config.coupling,
    };
  }

  private refreshSnapshot(): void {
    this.cachedSnapshot.spins = this.lattice.spinsArray();
    this.cachedSnapshot.dimensions = [...this.layout.dimensions];
    this.cachedSnapshot.rank = this.layout.definition.rank;
    this.cachedSnapshot.maxNeighbors = this.layout.maxNeighbors();
    this.cachedSnapshot.step = this.step;
    this.cachedSnapshot.geometry = this.layout.definition.name;
  }
}

File: ising-sim/backend/src/session.rs
use crate::config::SimConfig;
use crate::config::SimInitParams;
use crate::error::SimulationError;
use crate::geometry::{Geometry, LatticeGeometry};
use crate::interaction::NearestNeighbor;
use crate::lattice::Lattice;
use crate::metrics::{energy, magnetization};
use crate::monte_carlo::{SweepStats, sweep};
use rand::SeedableRng;
use rand::rngs::StdRng;

#[derive(Debug, Clone)]
pub struct SessionSnapshot {
    pub spins: Vec<i8>,
    pub width: usize,
    pub height: usize,
    pub step: u64,
    pub geometry: String,
}

#[derive(Debug, Clone, Copy)]
pub struct SessionMetrics {
    pub energy: f64,
    pub magnetization: f64,
    pub acceptance_rate: Option<f64>,
    pub temperature: f64,
    pub field: f64,
    pub coupling: f64,
}

#[derive(Debug)]
pub struct SimulationSession {
    geometry: LatticeGeometry,
    interaction: NearestNeighbor,
    lattice: Lattice,
    config: SimConfig,
    rng: StdRng,
    step: u64,
    running: bool,
    sweeps_per_tick: u64,
    last_acceptance_rate: Option<f64>,
}

impl SimulationSession {
    pub fn new(params: SimInitParams) -> Result<Self, SimulationError> {
        let config = params.into_config()?;
        let geometry = LatticeGeometry::from_name(&config.geometry, config.width, config.height)?;
        let interaction = NearestNeighbor::new(config.coupling);
        let lattice = Lattice::new(geometry.num_sites(), config.seed);
        let rng = match config.seed {
            Some(seed) => StdRng::seed_from_u64(seed.wrapping_add(1)),
            None => StdRng::from_os_rng(),
        };

        Ok(Self {
            geometry,
            interaction,
            lattice,
            config,
            rng,
            step: 0,
            running: false,
            sweeps_per_tick: 1,
            last_acceptance_rate: None,
        })
    }

    pub fn start(&mut self, steps_per_tick: Option<u64>) {
        self.sweeps_per_tick = steps_per_tick.unwrap_or(1).max(1);
        self.running = true;
    }

    pub fn pause(&mut self) {
        self.running = false;
    }

    pub fn is_running(&self) -> bool {
        self.running
    }

    pub fn sweeps_per_tick(&self) -> u64 {
        self.sweeps_per_tick
    }

    pub fn set_params(
        &mut self,
        temperature: Option<f64>,
        field: Option<f64>,
        coupling: Option<f64>,
    ) -> Result<(), SimulationError> {
        if let Some(t) = temperature {
            if t <= 0.0 {
                return Err(SimulationError::InvalidTemperature);
            }
            self.config.temperature = t;
        }
        if let Some(h) = field {
            self.config.field = h;
        }
        if let Some(j) = coupling {
            self.config.coupling = j;
            self.interaction = NearestNeighbor::new(j);
        }
        Ok(())
    }

    pub fn run_sweeps(&mut self, sweeps: u64) -> Result<SweepStats, SimulationError> {
        if sweeps == 0 {
            return Err(SimulationError::InvalidStep { sweeps });
        }

        let mut total = SweepStats::default();
        for _ in 0..sweeps {
            let stats = sweep(
                &mut self.lattice,
                &self.geometry,
                &self.interaction,
                &self.config,
                &mut self.rng,
            );
            total.attempts += stats.attempts;
            total.acceptances += stats.acceptances;
            self.step += 1;
        }
        self.last_acceptance_rate = Some(total.acceptance_rate());
        Ok(total)
    }

    pub fn snapshot(&self) -> SessionSnapshot {
        SessionSnapshot {
            spins: self.lattice.spins().to_vec(),
            width: self.config.width,
            height: self.config.height,
            step: self.step,
            geometry: self.geometry.name().to_string(),
        }
    }

    pub fn metrics(&self) -> SessionMetrics {
        SessionMetrics {
            energy: energy(
                &self.lattice,
                &self.geometry,
                &self.interaction,
                &self.config,
            ),
            magnetization: magnetization(&self.lattice),
            acceptance_rate: self.last_acceptance_rate,
            temperature: self.config.temperature,
            field: self.config.field,
            coupling: self.config.coupling,
        }
    }
}
